### PR TITLE
Sub tree evaluation

### PR DESCRIPTION
**Pull request by hetunandu**
## Description
Optimise evaluation by evaluating only a subset of all nodes needed to be evaluated.

Resolves: #1094 
Fixes: #2396

Co-authored by: @trishaanand 
Co-authored by: @ofpiyush

### CODE REVIEW

**Comment by ofpiyush** [comment]
I've made a poorly designed, but working sample app that breaks with subtree eval.

https://release.app.appsmith.com/applications/5fcf5deee297854451c7df47/pages/5fcf5deee297854451c7df49

vs

https://deploy-preview-1841--appsmith-editor-ce.netlify.app/applications/5fcf5deee297854451c7df47/pages/5fcf5deee297854451c7df49

**Comment by ofpiyush** [comment]
Same underlying problem with another case

https://release.app.appsmith.com/applications/5fcfb032bcf134084dbdd6f6/pages/5fcfb032bcf134084dbdd6f8

vs 
https://deploy-preview-1841--appsmith-editor-ce.netlify.app/applications/5fcfb032bcf134084dbdd6f6/pages/5fcfb032bcf134084dbdd6f8

File: app/client/src/widgets/TableWidget.tsx [hunk 1]
```diff
@@ -124,6 +127,13 @@ class TableWidget extends BaseWidget<TableWidgetProps, WidgetState> {
     };
   }
 
+  static getDerivedPropertiesMap(): DerivedPropertiesMap {
+    return {
+      selectedRow: `{{ _.get(this.filteredTableData, this.selectedRowIndex, _.mapValues(this.filteredTableData[0], () => undefined)) }}`,
+      selectedRows: `{{ this.filteredTableData.filter((item, i) => selectedRowIndices.includes(i) }); }}`,
```
Comment (ofpiyush) [inline app/client/src/widgets/TableWidget.tsx:133]:
```suggestion
      selectedRows: `{{ this.filteredTableData.filter((item, i) => this.selectedRowIndices && this.selectedRowIndices.includes(i)) }}`,
```

File: app/client/src/constants/ReduxActionConstants.tsx [hunk 2]
```diff
@@ -290,11 +289,13 @@ export const ReduxActionTypes: { [key: string]: string } = {
   CUT_SELECTED_WIDGET: "CUT_SELECTED_WIDGET",
   WIDGET_ADD_CHILDREN: "WIDGET_ADD_CHILDREN",
   SET_EVALUATED_TREE: "SET_EVALUATED_TREE",
+  SET_EVALUATION_DEPENDENCY_MAP: "SET_EVALUATION_DEPENDENCY_MAP",
   BATCH_UPDATES_SUCCESS: "BATCH_UPDATES_SUCCESS",
   UPDATE_CANVAS_STRUCTURE: "UPDATE_CANVAS_STRUCTURE",
   SET_SELECTED_WIDGET_ANCESTORY: "SET_SELECTED_WIDGET_ANCESTORY",
   START_EVALUATION: "START_EVALUATION",
   CURRENT_APPLICATION_NAME_UPDATE: "CURRENT_APPLICATION_NAME_UPDATE",
+  SET_WIDGET_LOADING: "SET_WIDGET_LOADING",
```
Comment (hetunandu) [inline app/client/src/constants/ReduxActionConstants.tsx:298]: Not needed anymore

**Pull request merged**